### ISSUE TITLE
[lx-brand] Systemd-228 mounts fail on NixOS-16.03

### ISSUE THREAD

**Issue by vizanto** [joyent/smartos-live#555]
```
[NOTICE: Zone rebooting]
booting system configuration /nix/store/f1i3grdzv3xygv96xmi2964h1rmyacp0-nixos-system-nixos-16.03pre75806.77f8f35
Failed to mount devtmpfs at /dev: No such device
Failed to mount tmpfs at /dev/shm: Device or resource busy
Failed to mount cgroup at /sys/fs/cgroup/systemd: Device or resource busy
[!!!!!!] Failed to mount API filesystems, freezing.
Freezing execution.
```

Steps to reproduce:
1. Create an lx-zone using http://datasets.at/ui/#!/configure/22ba59c6-8c49-11e5-87e5-7fbffbc05623
2. `nix-channel --add https://nixos.org/channels/nixos-unstable/`
3. `nixos-rebuild switch`  (this will fail)
4. Attach a console `vmadm console UUID`
5. Reboot the VM: `vmadm reboot UUID -F`

**Comment by pfmooney** [comment]
This situation may improve once [OS-4762](https://smartos.org/bugview/OS-4762) is resolved.

**Comment by jjelinek** [comment]
If you're on a system where you can run DTrace in the global zone, can you run this script while you're trying to boot the zone and post the output:

lx-syscall::mount:entry
{
    printf("%s -> %s(%s %s %s 0x%x %s)\n", execname, probefunc,
        copyinstr(arg0), copyinstr(arg1), copyinstr(arg2), arg3,
        (arg4 ? copyinstr(arg4) : "NULL"));
}

**Comment by vizanto** [comment]
i ran the script before and after updating:

Working boot:
```
dtrace: script 'script.d' matched 2 probes
CPU     ID                    FUNCTION:NAME
  9   3220                      mount:entry mount -> mount(none / none 0xc0ed0020 NULL)

  9   3220                      mount:entry mount -> mount(/nix/store /nix/store none 0xc0ed1000 NULL)

 10   3220                      mount:entry mount -> mount(/nix/store /nix/store none 0xc0ed1021 NULL)

 10   3220                      mount:entry mount -> mount(tmpfs /dev/shm tmpfs 0xc0ed0006 size=50%)

 10   3220                      mount:entry mount -> mount(tmpfs /run tmpfs 0xc0ed0000 mode=0755,size=25%)

 10   3220                      mount:entry mount -> mount(ramfs /run/keys ramfs 0xc0ed0000 NULL)

 10   3220                      mount:entry mount -> mount(tmpfs /var/setuid-wrappers tmpfs 0xc0ed0000 mode=0755)

  4   3220                      mount:entry mount -> mount(none /dev none 0xc0ed0020 size=5%)

  4   3220                      mount:entry mount -> mount(none /dev/shm none 0xc0ed0020 size=50%)

  4   3220                      mount:entry mount -> mount(none /run none 0xc0ed0020 size=25%)

  9   3220                      mount:entry systemd -> mount(sysfs /sys sysfs 0xe NULL)

  9   3220                      mount:entry systemd -> mount(devpts /dev/pts devpts 0xa mode=620,gid=3)

  9   3220                      mount:entry systemd -> mount(tmpfs /sys/fs/cgroup tmpfs 0x100000e mode=755)

  9   3220                      mount:entry systemd -> mount(cgroup /sys/fs/cgroup/systemd cgroup 0xe none,name=systemd,xattr)

  9   3220                      mount:entry systemd -> mount(tmpfs /sys/fs/cgroup tmpfs 0x100002f mode=755)

  3   3220                      mount:entry mount -> mount(swap /tmp tmpfs 0xc0ed0000 NULL)

  5   3220                      mount:entry systemd-logind -> mount(tmpfs /run/user/0 tmpfs 0x6 mode=0700,uid=0,gid=0,size=1717989376)
```

Broken boot (after update):
```
dtrace: script 'script.d' matched 2 probes
CPU     ID                    FUNCTION:NAME
  8   3220                      mount:entry mount -> mount(none / none 0xc0ed0020 NULL)

  8   3220                      mount:entry mount -> mount(/nix/store /nix/store none 0xc0ed1000 NULL)

  8   3220                      mount:entry mount -> mount(/nix/store /nix/store none 0xc0ed1021 NULL)

  8   3220                      mount:entry mount -> mount(tmpfs /dev/shm tmpfs 0xc0ed0006 size=50%)

  8   3220                      mount:entry mount -> mount(tmpfs /run tmpfs 0xc0ed0000 mode=0755,size=25%)

  8   3220                      mount:entry mount -> mount(ramfs /run/keys ramfs 0xc0ed0000 NULL)

  8   3220                      mount:entry mount -> mount(tmpfs /var/setuid-wrappers tmpfs 0xc0ed0000 mode=0755)

  9   3220                      mount:entry mount -> mount(none /dev none 0xc0ed0020 size=5%)

  9   3220                      mount:entry mount -> mount(none /dev/shm none 0xc0ed0020 size=50%)

  9   3220                      mount:entry mount -> mount(none /run none 0xc0ed0020 size=25%)

  7   3220                      mount:entry systemd -> mount(sysfs /sys sysfs 0xe NULL)

dtrace: error on enabled probe ID 1 (ID 3220: lx-syscall:sys64:mount:entry): invalid address (0x7fffff3e546f) in action #3 at DIF offset 12
  7   3220                      mount:entry systemd -> mount(devtmpfs /dev devtmpfs 0x1000002 mode=755)

dtrace: error on enabled probe ID 1 (ID 3220: lx-syscall:sys64:mount:entry): invalid address (0x7fffff3e4fff) in action #3 at DIF offset 12
  7   3220                      mount:entry systemd -> mount(devpts /dev/pts devpts 0xa mode=620,gid=3)

  7   3220                      mount:entry systemd -> mount(tmpfs /run tmpfs 0x1000006 mode=755)

  7   3220                      mount:entry systemd -> mount(tmpfs /sys/fs/cgroup tmpfs 0x100000e mode=755)

  7   3220                      mount:entry systemd -> mount(cgroup /sys/fs/cgroup/systemd cgroup 0xe none,name=systemd,xattr)

  7   3220                      mount:entry systemd -> mount(cgroup /sys/fs/cgroup/systemd cgroup 0xe none,name=systemd)

  7   3220                      mount:entry systemd -> mount(kdbusfs /sys/fs/kdbus kdbusfs 0xe NULL)
```

**Comment by pfmooney** [comment]
Is NixOS on a "real" kernel using kdbus?
It never made it into mainline Linux, AFAIK.

**Comment by vizanto** [comment]
Not sure how to investigate this, but these are the mounts on a real NixOS 16.03 machine:
```
# mount
proc on /proc type proc (rw,relatime)
sysfs on /sys type sysfs (rw,relatime)
devtmpfs on /dev type devtmpfs (rw,relatime,size=821844k,nr_inodes=2053360,mode=755)
tmpfs on /run type tmpfs (rw,relatime,size=4109212k,mode=755)
rpool/ROOT/nixos on / type zfs (rw,relatime,xattr,noacl)
rpool/ROOT/nixos on /nix/store type zfs (ro,relatime,xattr,noacl)
tmpfs on /dev/shm type tmpfs (rw,relatime,size=8218420k)
ramfs on /run/keys type ramfs (rw,relatime)
tmpfs on /var/setuid-wrappers type tmpfs (rw,relatime,mode=755)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=3,mode=620,ptmxmode=000)
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,mode=755)
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/run/current-system/systemd/lib/systemd/systemd-cgroups-agent,name=systemd)
cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpu,cpuacct)
cgroup on /sys/fs/cgroup/net_cls type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
mqueue on /dev/mqueue type mqueue (rw,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,relatime)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime)
/dev/md127 on /boot type ext4 (rw,relatime,data=ordered)
rpool/HOME on /home type zfs (rw,relatime,xattr,noacl)
tmpfs on /run/user/0 type tmpfs (rw,nosuid,nodev,relatime,size=1643684k,mode=700)
```

**Comment by vizanto** [comment]
And, no:
```
# ls /sys/fs/
cgroup  ext4
```

**Comment by pfmooney** [comment]
You may way to give this another try on a PI which includes the fix for OS-5155.  The symptoms are fairly similar (new-ish systemd freezing on boot).

**Comment by vizanto** [comment]
Something has changed all right:
```
[NOTICE: Zone rebooting]
booting system configuration /nix/store/ly0jsnzglcsz0wbaar6i87vq006hhv96-nixos-system-nixos-16.03pre76756.885acea
systemd 228 running in system mode. (+PAM +AUDIT -SELINUX +IMA +APPARMOR -SMACK -SYSVINIT +UTMP -LIBCRYPTSETUP +GCRYPT -GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD -IDN)
Detected virtualization container-other.
Detected architecture x86-64.

Welcome to NixOS 16.03pre76756.885acea (Emu)!

Assertion 'close_nointr(fd) != -EBADF' failed at src/basic/fd-util.c:67, function safe_close(). Aborting.
Caught <SEGV>, dumped core as pid 87423.
Freezing execution.
```

**Comment by vizanto** [comment]
@pfmooney :) ?

**Comment by vizanto** [comment]
Fixed in `SmartOS Live Image v0.147+ build: 20160304T005100Z`
